src/importer/sources/retroarch_source.py

```diff
@@ -0,0 +1,119 @@
+
+    def generator_builder(self) -> SourceIterationResult:
+        playlist_files = []
+        for file in os.listdir(self.source.config_location["playlists"]):
+            if file.endswith('.lpl'):
+                playlist_files.append(file)
+
+        playlist_items = []
+        for playlist_file in playlist_files:
+            open_file = open(str(self.source.config_location["playlists"]) + "/" + playlist_file)
+            try:
+                playlist_json = json.load(open_file)
+            except (JSONDecodeError, OSError, KeyError):
+                logging.warning("Cannot read playlist file: %s", str(playlist_file))
+                continue
+
+            for item in playlist_json["items"]:
+                # Select the core. Try the content's core first, then the playlist's
+                # default core.
+                core_path = item["core_path"]
+                if core_path == "DETECT":
+                    default_core = playlist_json["default_core_path"]
+                    if default_core:
+                        core_path = default_core
+                    else:
+                        logging.warning("Cannot find core for: %s", str(item["path"]))
+                        continue
+
+                # Build game
+                game_title = item["label"].split("(", 1)[0]
+                values = {
+                    "source": self.source.id,
+                    "added": int(time()),
+                    "name": game_title,
+                    "game_id": self.source.game_id_format.format(game_id=item["crc32"][:8]),
+                    "executable": self.source.executable_format.format(
+                        rom_path = item["path"],
+                        core_path = core_path,
+                    )
+                }
+
+                game = Game(values)
+                additional_data = {}
+
+                # Get boxart
+                boxart_image_name = item["label"].split(".", 1)[0] + ".png"
+                boxart_folder_name = playlist_file.split(".", 1)[0]
+                image_path = self.source.config_location["thumbnails"] / boxart_folder_name / "Named_Boxarts" / boxart_image_name
+                additional_data = {"local_image_path": image_path}
+
+                yield(game, additional_data)
+
+
+class RetroarchSource(Source):
+    args = ' -L "{core_path}" "{rom_path}"'
+
+    name = "Retroarch"
+    available_on = {"linux"}
+    iterator_class = RetroarchSourceIterator
+    executable_format = 'retroarch' + args
+
+    config_location = Location(
+        schema_key="retroarch-location",
+        candidates=(
+            shared.flatpak_dir / "org.libretro.RetroArch" / "config" / "retroarch",
+            shared.config_dir / "retroarch",
+            shared.home / ".config" / "retroarch",
+        ),
+        paths={
+            "playlists": (True, "playlists"),
+            "thumbnails": (True, "thumbnails"),
+        },
+    )
+
+    # Check if installation is flatpak'd.
```

Comment:
This should be a property, the source location should not be resolved from the candidate paths until needed:

```py
@property
def executable_format(self):
        self.config_location.resolve()
        is_flatpak = self.config_location.root.is_relative_to(shared.flatpak_dir)
        base = "flatpak run org.libretro.RetroArch" if is_flatpak else "retroarch"
        args = '-L "{core_path}" "{rom_path}"'
        return f"{base} {args}"
```

You can then remove the related unnecessary code in the class definition

Reply:
@kra-mo is that a better option than having a `native` variant and a `flatpak` variant with different location candidates? I guess you would prefer this, because it doesn't create a source variant, but I want your input just to be sure for other emulation sources.

Reply:
If the executable format differs, I'm fine with having two different sources.

Reply:
So should I go ahead and add this property? Just as final confirmation--
I'm slowly going through everything and realising that I might have taken on a bit much for my first open source pull request :sweat_smile: I'll see if I can manage though!

Reply:
> So should I go ahead and add this property?

yes haha, sorry for the confusion

Reply:
Hey don't worry, you're doing great :)

Reply:
Just a heads up, there is still `args` at the top of the source